### PR TITLE
Improve diagnostics when tests fail due to an asynchronous operation timeout

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

**Comment by sharwell** [comment]
@Pilchie for approval. Code touches production assemblies but is only enabled in test scenarios.

File: src/Workspaces/Core/Portable/Shared/TestHooks/IAsynchronousOperationListenerProvider.cs [hunk 1]
```diff
@@ -61,11 +61,20 @@ internal sealed class AsynchronousOperationListenerProvider : IAsynchronousOpera
...
 
+        /// <summary>
+        /// Provides a default value for <see cref="_enableDiagnosticTokens"/>.
+        /// </summary>
+        private static bool? s_enableDiagnosticTokens;
+
         public static void Enable(bool enable)
+            => Enable(enable, diagnostics: false);
+
+        public static void Enable(bool enable, bool diagnostics)
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Shared/TestHooks/IAsynchronousOperationListenerProvider.cs]: ❗️ This should be `bool?`

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs [hunk 2]
```diff
@@ -226,7 +226,7 @@ private bool UpToDate
             }
 
             public void RegisterNotification(Action action, int delay, CancellationToken cancellationToken)
-                => _notificationService.RegisterNotification(action, delay, _asyncListener.BeginAsyncOperation("TagSource"), cancellationToken);
+                => _notificationService.RegisterNotification(action, delay, _asyncListener.BeginAsyncOperation(typeof(TTag).Name), cancellationToken);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs]: Nit: want to do something like typeof(...).Name + "TagSource"? That way if you have a greppable string in addition to useful detail. (If you ended up holding a token with name IOutliningTag, it's less obvious what to find?)

File: src/Workspaces/Core/Portable/Shared/TestHooks/IAsynchronousOperationListenerProvider.cs [hunk 3]
```diff
@@ -192,9 +201,16 @@ private bool DiagnosticTokensEnabled
...
-                    // if _trackingBehavior has never been set, check environment variable to see whether it should be enabled.
-                    var enabled = Environment.GetEnvironmentVariable("RoslynWaiterDiagnosticTokenEnabled");
-                    _enableDiagnosticTokens = string.Equals(enabled, "1", StringComparison.OrdinalIgnoreCase) || string.Equals(enabled, "True", StringComparison.OrdinalIgnoreCase);
+                    if (s_enableDiagnosticTokens.HasValue)
+                    {
+                        _enableDiagnosticTokens = s_enableDiagnosticTokens;
+                    }
+                    else
+                    {
+                        // if _trackingBehavior has never been set, check environment variable to see whether it should be enabled.
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/TestHooks/IAsynchronousOperationListenerProvider.cs]: Not your fault, but comment referencing stale variable, probably.

File: src/Workspaces/CoreTestUtilities/UseExportProviderAttribute.cs [hunk 4]
```diff
@@ -112,7 +113,14 @@ public override void After(MethodInfo methodUnderTest)
...
                         catch (OperationCanceledException ex) when (timeoutTokenSource.IsCancellationRequested)
                         {
-                            throw new TimeoutException("Failed to clean up listeners in a timely manner.", ex);
+                            var messageBuilder = new StringBuilder("Failed to clean up listeners in a timely manner.");
+                            foreach (var token in ((AsynchronousOperationListenerProvider)listenerProvider).GetTokens())
+                            {
+                                messageBuilder.AppendLine().Append($"  {token}");
+                            }
+
+                            Debugger.Launch();
```
Comment (jasonmalinowski) [inline src/Workspaces/CoreTestUtilities/UseExportProviderAttribute.cs]: I think we normally call our fail fast helper in tests. I could imagine that if procdump is attached that counts as a debugger, which will make this call do nothing per the docs.

Comment (jasonmalinowski) [reply src/Workspaces/CoreTestUtilities/UseExportProviderAttribute.cs]: Also, if I'm running from command line, I may not want a JIT debug window popping up. (If you do, congrats, you get to rewrite our helpers to do that uniformly. :smile:)

Comment (sharwell) [reply src/Workspaces/CoreTestUtilities/UseExportProviderAttribute.cs]: Unintentional holdover from local debugging

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs]: :arrow_right: Name chosen to provide meaning without incurring a fast path allocation (the method is a NOP in production)

**Comment by sharwell** [comment]
Created #25908 to help diagnose the test failure here.

**Pull request merged**